feat: Add Metrics V3 API

donatello opened the pull request:
## Description

Metrics V3 is mainly a reorganization of metrics into smaller groups of
metrics and the removal of internal aggregation of metrics received from
peer nodes in a MinIO cluster.

This change adds the endpoint `/minio/metrics/v3` as the top-level metrics
endpoint and under this, various sub-endpoints are implemented. These
are currently documented in `docs/metrics/v3.md`

The handler will serve metrics at any path
`/minio/metrics/v3/PATH`, as follows:

  when PATH is a sub-endpoint for a metrics group => serves the group of
  metrics under that path; or

  when PATH is a (non-empty) parent directory of the sub-endpoints
  listed above => serves metrics from each child sub-endpoint of PATH.

  otherwise, returns a no resource found error


All available metrics are listed in the `docs/metrics/v3.md`. More will
be added subsequently.

## Motivation and Context

Reworking of metrics for better observability

## How to test this PR?



## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

harshavardhana commented on cmd/metrics-v3-handler.go:
```diff
@@ -0,0 +1,167 @@
+	nodeAPIS3CollectorPath            = "/node/api/s3"
+	nodeAPIBucketS3CollectorPath      = "/node/api/bucket/s3"
+	nodeNetworkInternodeCollectorPath = "/node/network/internode"
+	nodeDebugCollectorPath            = "/node/debug"
```
We must always prefix with `/minio` since that's the disallowed bucket. `/node` will conflict with an existing bucket.

donatello replied:
These are subpaths under the reserved bucket

donatello committed:
feat: Add Metrics V3 API

Metrics V3 is mainly a reorganization of metrics into smaller groups of
metrics and the removal of internal aggregation of metrics received from
peer nodes in a MinIO cluster.

This change adds the endpoint `/minio/metrics/v3` as the top-level metrics
endpoint and under this, various sub-endpoints are implemented. These
are currently documented in `docs/metrics/v3.md`

The handler will serve metrics at any path
`/minio/metrics/v3/PATH`, as follows:

  when PATH is a sub-endpoint listed above => serves the group of
  metrics under that path; or

  when PATH is a (non-empty) parent directory of the sub-endpoints
  listed above => serves metrics from each child sub-endpoint of PATH.

  otherwise, returns a no resource found error

All available metrics are listed in the `docs/metrics/v3.md`. More will
be added subsequently.

donatello committed:
Rename Metric to MetricV2 and

- MetricType to MetricTypeV2
- MetricsGroup to MetricsGroupV2

Update generated code after renames.

poornas commented on cmd/metrics-v3-cluster-usage.go:
```diff
@@ -0,0 +1,188 @@
+	dataUsageInfo, err := loadDataUsageFromBackend(ctx, objLayer)
+	if err != nil {
+		logger.LogIf(ctx, err)
+		return nil
```
with the break up of metrics info across endpoints, there is redundant calls to load usage/ fetch storage info - perhaps it may be ideal to cache data for related metric endpoints once as opposed to caching per metric endpoint

harshavardhana replied:
We can start heavily using `cachevalue.Cache` implementations here, we really don't need such frequent loads from disk.

donatello replied:
This is done now, a separate structure holds all the cached metrics. Not everything is cached.

harshavardhana commented on cmd/metrics-router.go:
```diff
@@ -48,15 +60,24 @@ const (
 	metricsRouter.Handle(prometheusMetricsV2ResourcePath, auth(metricsResourceHandler()))
+
+	if env.Get(EnvMetricsV3FeatureGate, "off") != "off" {
+		fmt.Printf("Metrics v3 feature gate is enabled\n")
```
```suggestion
		logger.Info("Metrics v3 feature gate is enabled\n")
```

shtripat commented on cmd/metrics-v3-handler.go:
```diff
@@ -0,0 +1,259 @@
+	matchingMG := make(map[collectorPath]*MetricsGroup)
+	for _, collPath := range h.metricsData.collectorPaths {
+		if collPath.isDescendantOf(path) {
+			fmt.Printf("Listing metrics under %s\n", collPath)
```
do we need this debug print?

shtripat commented on cmd/metrics-v3-handler.go:
```diff
@@ -0,0 +1,259 @@
+		http.Error(w, "Metrics Resource Not found", http.StatusNotFound)
+	})
+
+	fmt.Printf("Requested metric path: %s (listOnly=%v) (buckets=%v)\n", path,
```
same as above

shtripat commented on cmd/metrics-v3-handler.go:
```diff
@@ -0,0 +1,259 @@
+	var gatherers []prometheus.Gatherer
+	for _, collectorPath := range h.metricsData.collectorPaths {
+		if collectorPath.isDescendantOf(path) {
+			fmt.Printf("Serving collector path %s\n", collectorPath)
```
remove if not needed

shtripat commented on cmd/metrics-v3-system-network.go:
```diff
@@ -0,0 +1,61 @@
+	internodeDialedErrorsTotalMD = NewCounterMD(internodeDialErrorsTotal,
+		"Total number of internode TCP dial timeouts and errors")
+	internodeDialAvgTimeNanosMD = NewGaugeMD(internodeDialAvgTimeNanos,
+		"Average dial time of internodes TCP calls in nanoseconds")
```
```suggestion
		"Average dial time of internode TCP calls in nanoseconds")
```

shtripat commented on cmd/metrics-v3-types.go:
```diff
@@ -0,0 +1,487 @@
+	Help           string
+	VariableLabels []string
+
+	// managed values follow:
```
bogus comment?

donatello replied:
`labelSet` is a managed value - meaning it is not initialized when creating the struct, but is used via the `getLabelSet` method.

shtripat reviewed (approved):
Looks good. Few minor ones and will verify the metrics listing.

shtripat commented on cmd/metrics-v3-cluster-usage.go:
```diff
@@ -0,0 +1,181 @@
+	usageVersionsTotal                  = "versions_total"
+	usageDeleteMarkersTotal             = "delete_markers_total"
+	usageBucketsTotal                   = "buckets_total"
+	usageObjectSizeDistribution         = "object_size_distribution"
```
Here metric name becomes `minio_cluster_usage_objects_object_size_distribution` can we just have like `minio_cluster_usage_objects_size_distribution`

shtripat commented on cmd/metrics-v3-cluster-usage.go:
```diff
@@ -0,0 +1,181 @@
+	usageDeleteMarkersTotal             = "delete_markers_total"
+	usageBucketsTotal                   = "buckets_total"
+	usageObjectSizeDistribution         = "object_size_distribution"
+	usageObjectVersionCountDistribution = "object_version_count_distribution"
```
Can we have like `minio_cluster_usage_objects_version_count_distribution`? currently it looks like `minio_cluster_usage_objects_object_version_count_distribution`

shtripat commented on cmd/metrics-v3-cluster-usage.go:
```diff
@@ -0,0 +1,181 @@
+const (
+	usageSinceLastUpdateSeconds         = "since_last_update_seconds"
+	usageTotalBytes                     = "total_bytes"
+	usageObjectsTotal                   = "objects_total"
```
Can we have like `minio_cluster_usage_objects_total` ? currently it looks like `minio_cluster_usage_objects_objects_total`

shtripat commented on cmd/metrics-v3-cluster-usage.go:
```diff
@@ -0,0 +1,181 @@
+)
+
+// loadClusterUsageBucketMetrics - `MetricsLoaderFn` to load bucket usage metrics.
+func loadClusterUsageBucketMetrics(ctx context.Context, m MetricValues, c *metricsCache, buckets []string) error {
```
`/minio/metrics/v3/cluster/usage/buckets` would always work with query param `?buckets=name1,name2`.
I think this is intentional.

donatello replied:
Yes, to get metrics about buckets, specific buckets must be provided via the query parameter. This is also documented in `v3.md`.

shtripat reviewed (approved):
LGTM

harshavardhana commented on cmd/server-main.go:
```diff
@@ -341,6 +341,7 @@ func serverHandleCmdArgs(ctxt serverCtxt) {
 	globalLocalNodeName = GetLocalPeer(globalEndpoints, globalMinioHost, globalMinioPort)
 	nodeNameSum := sha256.Sum256([]byte(globalLocalNodeName))
 	globalLocalNodeNameHex = hex.EncodeToString(nodeNameSum[:])
+	globalLocalPoolIdx = getLocalNodePoolIdx(globalLocalNodeName)
```
I would suggest that this is returned via the `GetLocalPeer()` function itself to avoid doing double the work.

harshavardhana replied:
actually this code `getLocalProperty()` does not set the PoolNumber so it is not really working properly.

harshavardhana replied:
I am sending a correct fix.

harshavardhana replied:
also a single server can belong to two pools at the same time.

donatello replied:
I remove this global now, we don't need it in this form anyway.
